### DISCUSSION TITLE
Remove-PnPFileVersion access denied

### DISCUSSION THREAD

**Discussion by OmarSeri** [pnp/powershell#3130] [Q&A]
Trying to remove Sharepoint library files versions with Remove-PnPFileVersion I've got Access denied error.
I cannot find out what's going on since I can delete file versions thought GUI.

**Comment by AlexSen** [comment]
Same issue here

**Reply by kasperbolarsen** [reply]
I added a sample using Remove-PnPFileVersion to the PnP Script Samples, and it worked last week, when I checked it last.
Perhaps it can help identifying the issue?

https://pnp.github.io/script-samples/spo-file-version-trimmer/README.html?tabs=pnpps

**Reply by kwangj** [reply]
Hi.

I use below script:
-----------------------------
$PSLogin = @{
    Tenant = "XXXXXX"
    ClientId = "XXXXXX"
    Thumbprint = "XXXXXX"
}
 Connect-PnPOnline -Url $AdminSiteURL -ClientId $PSLogin.ClientId  -Tenant $PSLogin.Tenant -Thumbprint  $PSLogin.Thumbprint   

And  I grant App permission "SharePoint - Sites.FullControl.All"

 and delete fileversion by Remove-PnPFileVersion result is  "Access denied"

What permission is need to execute "Remove-PnPFileVersion"?

**Comment by waves-rider** [comment]
Same issue here with the following error ; 

⚠️ Protected: [FILE_NAME].pptx - Versions of this item cannot be deleted because they are locked, or a retention policy is applied.

Even though, there is no retention policy applied neither on the site or in purview.

**Reply by toffitomek** [reply]
came here to report similar problem, I can't even see file version, only get:
`Get-PnPFileVersion: Access denied.
`

**Reply by waves-rider** [reply]
Are you using an entra app ?
With full control permission you shouldn't have this permission issue.

**Reply by IsaacM72** [reply]
I don't want to hi-jack the thread but I'm running into the same issue. Entra app was setup, initially with Read and write permissions, which did not work. Changed to Fullsite control, also not working. User account is SP admin, even added to the as Owner and Member to the site. Script runs fine showing versions but gives Access denied error when running the following section:
Remove-PnPFileVersion -Url $FileRelativePath -Identity $Versions[$i].ID -Force

Will start a new thread in any case as not sure if OP is still looking at this

**Reply by AndersRask** [reply]
> Same issue here with the following error ;
> 
> ⚠️ Protected: [FILE_NAME].pptx - Versions of this item cannot be deleted because they are locked, or a retention policy is applied.
> 
> Even though, there is no retention policy applied neither on the site or in purview.


This could be because the file is locked by owner or that there is no checked in version. Check the version history of the file to see if file was ever checked in or is in draft.

More on how to manage files with no checked in version
https://www.sharepointdiary.com/2017/07/powershell-to-manage-files-which-nave-no-checked-in-version.html

For locked files you can check some samples here
https://github.com/pnp/script-samples/blob/main/scripts/spo-record-lock-unlock-file/README.md/

**Reply by IsaacM72** [reply]
@waves-rider I assume you can delete the versions manually under the Storage Metrics section?

**Reply by waves-rider** [reply]
@IsaacM72 yes I can.

My solution is to set each site settings to not do versioning anymore, but I don't recommend doing this.

**Comment by kasperbolarsen** [comment]
this one works with 3.1.0

`function DeleteFileVersions {
    param (
        [Parameter(Mandatory=$true)]
        [string]$Path,
        [Parameter(Mandatory=$true)]
        [string]$siteUrl
        ,
        [Parameter(Mandatory=$true)]
        [string]$remainingVersions
    )
    try
    {
        
        Connect-PnPOnline -Url $siteUrl -ClientId $PnPClientId -Interactive 
        $Path = $Path.Replace($TenantURL,"")
        $file = Get-PnPFile  -Url $Path -AsFileObject -ErrorAction Stop
    }
    catch
    {
        write-host -f Red "`tError:" $_.Exception.Message            
    }

    if(-not $file)
    {
        continue
    }
    $versions = Get-PnPProperty  -ClientObject $file -Property Versions 
    $versions = $versions | Sort-Object -Property Created 
    $numberOfVersionsToDelete = $versions.Count - $remainingVersions
    $counter = 0
    foreach($VersionToDelete in $versions) 
    {
        if($counter -ge $numberOfVersionsToDelete)
        {
            break
        }
        $versionNumber = [int]$VersionToDelete.VersionLabel
        write-host "Deleting version $($versionNumber) of file $($Path)"
        Remove-PnPFileVersion -Url $Path -Identity $VersionToDelete.Id –Force -Connection $siteconn            
        $counter++
    }
    #return $TotalFileSize
}
`

**Reply by GerhardSauer** [reply]
does not work for me.
Still get the access denied...
for get-pnpfileversion and remove-pnpfileversion.

Reading the file version via Get-PnPProperty works.
